src/features/Folding.ts

```diff
@@ -0,0 +1,415 @@
+        this.matchScopeElements(tokens,
+                                "punctuation.section.group.begin.powershell",
+                                "punctuation.section.group.end.powershell",
+                                MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find contiguous here strings   @' -> '@
+        this.matchContiguousScopeElements(tokens,
+                                          "string.quoted.single.heredoc.powershell",
+                                          MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find contiguous here strings   @" -> "@
+        this.matchContiguousScopeElements(tokens,
+                                          "string.quoted.double.heredoc.powershell",
+                                          MatchType.Region, document)
+            .forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching Braces   #region -> #endregion
+        this.matchScopeElements(this.extractRegionScopeElements(tokens, document),
+                                "custom.start.region",
+                                "custom.end.region",
+                                MatchType.Region, document)
+            .forEach( (x) => { matchedTokens.push(x); });
+
+        // Find blocks of line comments   # comment1\n# comment2\n...
+        this.matchBlockCommentScopeElements(tokens, document).forEach((x) => { matchedTokens.push(x); });
+
+        // Find matching function definitions   <# -> #>
+        this.matchScopeElements(tokens,
+                                "punctuation.definition.comment.block.begin.powershell",
+                                "punctuation.definition.comment.block.end.powershell",
+                                MatchType.Comment, document)
+            .forEach( (x) => { matchedTokens.push(x); });
+
+        return matchedTokens;
+    }
+
+    // Search all of the loaded extenions for the PowerShell grammar file
+    private powerShellGrammarPath(): string {
+        try {
+            const psGrammars =
+                vscode.extensions.all
+                    .filter((x) => x.packageJSON && x.packageJSON.contributes && x.packageJSON.contributes.grammars)
+                    .reduce((a: IExtensionGrammar[], b) =>
+                        [...a, ...(b.packageJSON as IExtensionPackage).contributes.grammars
+                            .map((x) => Object.assign({ extensionPath: b.extensionPath }, x))], [])
+                    .filter((x) => x.language === "powershell");
+            if (psGrammars.length === 0) { return ""; }
+            return path.join(psGrammars[0].extensionPath, psGrammars[0].path);
+        } catch (err) { return ""; }
+    }
+
+    // Returns the PowerShell grammar parser
+    private grammar() {
+        if (this.powershellGrammar !== undefined) { return this.powershellGrammar; }
+        const registry = new tm.Registry();
```

Comment:
Yup.  Came to same conclusion too.  I moved the grammar loading from the FoldingProvider, into the FoldingFeature.  This made more sense and should also make testing easier  I made the grammar loading far more defensive, and if it fails, it will not register the folding provider, so at least you'll get the default folding from VSCode.